spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -68,6 +70,37 @@ case class RewriteMergeInto(spark: SparkSession) extends Rule[LogicalPlan] with
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
     plan resolveOperators {
+      case MergeIntoTable(target: DataSourceV2Relation, source, cond, matchedActions, notMatchedActions)
+          if matchedActions.isEmpty && notMatchedActions.size == 1 && isIcebergRelation(target) =>
+
+        val targetTableScan = buildSimpleScanPlan(target, cond)
+
+        val insertAction = notMatchedActions match {
+          case Seq(action: InsertAction) =>
+            action
+          case _ =>
+            throw new AnalysisException("Only insert actions are supported in NOT MATCHED clauses")
```

Comment:
Isn't this guaranteed by our analyzer rule?

Reply:
Yep, removed.